app/gui/src/presenter/graph/widgets.rs

```diff
@@ -0,0 +1,352 @@
+        retain_node_expressions(ViewNodeId, HashSet<ast::Id>),
+        /// Remove all widget queries of given node.
+        remove_node(ViewNodeId),
+    }
+    Output {
+        /// Emitted when the node's visualization has been set.
+        widget_data(ViewNodeId, WidgetUpdates),
+    }
+}
+
+/// Graph widgets controller. One requester per function call. Responsible for querying the language
+/// server for information about the node's widget metadata or resolving from local cache.
+#[derive(Debug)]
+pub struct Widgets {
+    frp:   Frp,
+    #[allow(dead_code)]
+    model: Rc<RefCell<Model>>,
+}
+
+impl Deref for Widgets {
+    type Target = Frp;
+
+    fn deref(&self) -> &Self::Target {
+        &self.frp
+    }
+}
+
+impl Widgets {
+    /// Constructor
+    pub fn new(executed_graph: ExecutedGraph) -> Self {
+        let (manager, manager_notifications) = Manager::new(executed_graph.clone_ref());
+        let frp = Frp::new();
+
+        let model = Rc::new(RefCell::new(Model {
+            manager,
+            graph: executed_graph.clone_ref(),
+            widgets_of_node: default(),
+            widget_queries: default(),
+        }));
+
+        let network = &frp.network;
+        let input = &frp.input;
+        let output = &frp.private.output;
+
+        frp::extend! { network
+            updates_from_cache <- input.request_widgets.filter_map(
+                f!((definition) model.borrow_mut().request_widget(definition))
+            );
+            output.widget_data <+ updates_from_cache;
+            eval input.retain_node_expressions(((node_id, expr_ids)) {
+                model.borrow_mut().retain_node_expressions(*node_id, expr_ids)
+            });
+            eval input.remove_node((node_id) model.borrow_mut().remove_node(*node_id));
+        };
+
+        let out_widget_data = output.widget_data.clone_ref();
+        let weak = Rc::downgrade(&model);
+        spawn_stream_handler(weak, manager_notifications, move |notification, model| {
+            let data = model.borrow_mut().handle_notification(notification);
+            if let Some(data) = data {
+                out_widget_data.emit(data);
+            }
+            std::future::ready(())
+        });
+
+        Self { frp, model }
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+/// Model of the Widgets controller. Manages the widget queries.
+#[derive(Debug)]
+pub struct Model {
+    manager:         Rc<Manager>,
+    graph:           ExecutedGraph,
+    widgets_of_node: HashMap<ViewNodeId, Vec<ast::Id>>,
+    widget_queries:  HashMap<ast::Id, QueryData>,
+}
+
+/// Data of the query that
+#[derive(Debug)]
+struct QueryData {
+    node_id:         ViewNodeId,
+    call_expression: ast::Id,
+    method_name:     ImString,
+    arguments:       Vec<ImString>,
+    last_update:     Option<WidgetUpdates>,
+}
+
+/// Definition of a widget request. Defines the operation expression that the widgets will be
+/// attached to.
```

Comment:
Could you expland what "Defines the operation expression that the widgets will be attached to." means, please?